### PR TITLE
Support end to end encryption for node client

### PR DESCRIPTION
**Pull request by tuteng**
After this pr is https://github.com/apache/pulsar/pull/9074 merged, we can consider enabling end-to-end encryption for the node client.

* Enabling end-to-end encryption
* Add producer, consumer examples and test certificates

### CODE REVIEW

**Comment by massakam** [comment]
The milestone for https://github.com/apache/pulsar/pull/9074 is 2.7.1 and has not yet been released. We cannot merge this PR until version 2.7.1 of the C/C++ client library is released.

File: examples/encryption-consumer.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+  const consumer = await client.subscribe({
+    topic: 'persistent://public/default/my-topic',
+    subscription: 'sub1',
+    subscriptionType: 'Shared',
+    ackTimeoutMs: 10000,
+    publicKeyPath: "./certificate/public-key.client-rsa.pem",
+    privateKeyPath: "./certificate/private-key.client-rsa.pem"
+  });
+
+  console.log(consumer)
```
Comment (massakam) [inline examples/encryption-consumer.js:39]: It seems that no useful information is output.
```
Consumer {}
```

File: examples/encryption-producer.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  const producer = await client.createProducer({
+    topic: 'persistent://public/default/my-topic',
+    sendTimeoutMs: 30000,
+    batchingEnabled: true,
+    publicKeyPath: "./certificate/public-key.client-rsa.pem",
+    privateKeyPath: "./certificate/private-key.client-rsa.pem",
+    encryptionKey: "encryption-key"
+  });
+
+  console.log(producer.ProducerConfig)
```
Comment (massakam) [inline examples/encryption-producer.js:39]: `producer.ProducerConfig` is undefined.

File: src/ConsumerConfig.cc [hunk 3]
```diff
@@ -163,6 +172,23 @@ ConsumerConfig::ConsumerConfig(const Napi::Object &consumerConfig,
       pulsar_consumer_set_read_compacted(this->cConsumerConfig, 1);
     }
   }
+
+  if ((consumerConfig.Has(CFG_PUBLIC_KEY_PATH) && consumerConfig.Has(CFG_PRIVATE_KEY_PATH)) &&
```
Comment (massakam) [inline src/ConsumerConfig.cc:176]: Is the public key required on the consumer side?

File: src/ProducerConfig.cc [hunk 4]
```diff
@@ -153,6 +162,25 @@ ProducerConfig::ProducerConfig(const Napi::Object& producerConfig) : topic("") {
                                                  value.Utf8Value().c_str());
     }
   }
+
+  if ((producerConfig.Has(CFG_PUBLIC_KEY_PATH) && producerConfig.Has(CFG_PRIVATE_KEY_PATH)) &&
```
Comment (massakam) [inline src/ProducerConfig.cc:166]: On the producer side, a private key does not seem to be required.

File: src/ProducerConfig.cc [hunk 5]
```diff
@@ -153,6 +162,25 @@ ProducerConfig::ProducerConfig(const Napi::Object& producerConfig) : topic("") {
...
+                                                                privateKeyPath.c_str());
+    if (producerConfig.Has(CFG_ENCRYPTION_KEY) && producerConfig.Get(CFG_ENCRYPTION_KEY).IsString()) {
+      std::string encryptionKey = producerConfig.Get(CFG_ENCRYPTION_KEY).ToString().Utf8Value();
+      pulsar_producer_configuration_set_encryption_key(this->cProducerConfig, encryptionKey.c_str());
+    }
+    if (producerConfig.Has(CFG_CRYPTO_FAILURE_ACTION) &&
+        producerConfig.Get(CFG_CRYPTO_FAILURE_ACTION).IsString()) {
+      std::string cryptoFailureAction = producerConfig.Get(CFG_CRYPTO_FAILURE_ACTION).ToString().Utf8Value();
+      pulsar_producer_configuration_set_crypto_failure_action(
+          this->cProducerConfig, PRODUCER_CRYPTO_FAILURE_ACTION.at(cryptoFailureAction));
```
Comment (massakam) [inline src/ProducerConfig.cc:181]: We should check if `cryptoFailureAction` is contained in `PRODUCER_CRYPTO_FAILURE_ACTION`.

File: tests/end_to_end.test.js [hunk 6]
```diff
@@ -665,5 +669,87 @@ const Pulsar = require('../index.js');
...
+        encryptionKey: 'encryption-key',
+      });
+
+      const consumer = await client.subscribe({
+        topic,
+        subscription: 'sub1',
+        subscriptionType: 'Shared',
+        ackTimeoutMs: 10000,
+        publicKeyPath: './public-rsa.pem',
+        privateKeyPath: './private-rsa.pem',
```
Comment (massakam) [inline tests/end_to_end.test.js:694]: Are these paths correct?

File: tests/end_to_end.test.js [hunk 7]
```diff
@@ -665,5 +669,87 @@ const Pulsar = require('../index.js');
...
+        consumer.acknowledge(msg);
+        results.push(msg.getData().toString());
+      }
+      expect(lodash.difference(messages, results)).toEqual([]);
+      await producer.close();
+      await consumer.close();
+      await client.close();
+    });
+
+    test('Failed produce and consume encryption', async () => {
```
Comment (massakam) [inline tests/end_to_end.test.js:719]: Is this test case necessary? It doesn't seem to be a test related to encryption.

**Comment by tuteng** [comment]
Now that 2.7.1 has been released, this feature is available, I fixed your comment, PTAL @massakam Thanks

**Comment by tuteng** [comment]
The test failed due to a deb package issue, which will be fixed later

**Comment by tuteng** [comment]
@massakam PTAL

**Review by massakam — APPROVED** [review]
LGTM

**Pull request merged**